streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -284,18 +286,23 @@ public void resume() {
     }
 
     /**
-     * An active task is processable if its buffer contains data for all of its input source topic partitions
+     * An active task is processable if its buffer contains data for all of its input
+     * source topic partitions, or if it is enforced to be processable
      */
     public boolean isProcessable() {
+        return enforcedProcess || partitionGroup.allPartitionsBuffered();
+    }
+
+    public void maybeEnforceProcess(final long now) {
         if (partitionGroup.allPartitionsBuffered()) {
-            return true;
-        } else if (partitionGroup.numBuffered() > 0 && --waits < 0) {
+            enforcedProcess = false;
+        } else if (partitionGroup.numBuffered() > 0 && now - lastEnforcedProcess > maxTaskIdleMs) {
             taskMetrics.taskEnforcedProcessSensor.record();
-            waits = WAIT_ON_PARTIAL_INPUT;
-            return true;
+            lastEnforcedProcess = now;
```

Comment:
That's a good catch, but if we move `isProcessable()` inside the iteration, it will mean that we will only enforce-process one record for every `max.idle.ms` right?

My original thought is that once we've decided to enforce process, we'll enforce for that whole thread iteration.